[Bug] GLM-5.2 crash with pipeline-parallel

### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

When using `sglang serve --model-path /data/hub/models--lukealonso--GLM-5.2-NVFP4/snapshots/8a1f4a13204acf2b7ac840375efaed64c231c522 --host 0.0.0.0 --port 8000 --served-model-name glm-5.2 --pipeline-parallel-size 2 --max-total-tokens 300000 --kv-cache-dtype bfloat16 --disable-shared-experts-fusion --enable-metrics --mem-fraction-static 0.9 --sleep-on-idle --trust-remote-code --quantization modelopt_fp4 --reasoning-parser glm45 --tool-call-parser glm47 --log-requests --log-requests-level 0 --log-requests-format text`, it crashes with:

```
2026-06-17T19:49:56.097780059+02:00 stderr F [2026-06-17 17:49:56] User lacks permission to set NUMA affinity, skipping NUMA node configuration for GPU. If using docker, try adding --cap-add SYS_NICE to your docker run command.
2026-06-17T19:49:56.099728872+02:00 stderr F [2026-06-17 17:49:56] User lacks permission to set NUMA affinity, skipping NUMA node configuration for GPU. If using docker, try adding --cap-add SYS_NICE to your docker run command.
2026-06-17T19:50:01.564381139+02:00 stderr F [2026-06-17 17:50:01] Tokenizer for /data/hub/models--lukealonso--GLM-5.2-NVFP4/snapshots/8a1f4a13204acf2b7ac840375efaed64c231c522 is still TokenizersBackend after retries with --trust-remote-code. Model-specific tokenizer attributes may be missing.
2026-06-17T19:50:02.272155824+02:00 stderr F [2026-06-17 17:50:02] Using default HuggingFace chat template with detected content format: openai
2026-06-17T19:50:02.401257296+02:00 stderr F [2026-06-17 17:50:02] Auto-detected template features: reasoning_config=ReasoningToggleConfig(toggle_param='enable_thinking', default_enabled=True, special_case=None), reasoning_parser=glm45, tool_call_parser=glm45
2026-06-17T19:50:06.752723029+02:00 stderr F /usr/local/lib/python3.12/dist-packages/torchao/quantization/quant_api.py:1731: SyntaxWarning: invalid escape sequence '\.'
2026-06-17T19:50:06.752750137+02:00 stderr F   """Configuration class for applying different quantization configs to modules or parameters based on their fully qualified names (FQNs).
2026-06-17T19:50:08.249889859+02:00 stderr F /usr/local/lib/python3.12/dist-packages/torchao/quantization/quant_api.py:1731: SyntaxWarning: invalid escape sequence '\.'
2026-06-17T19:50:08.249913328+02:00 stderr F   """Configuration class for applying different quantization configs to modules or parameters based on their fully qualified names (FQNs).
2026-06-17T19:50:10.862362644+02:00 stderr F /usr/local/lib/python3.12/dist-packages/torchao/quantization/quant_api.py:1731: SyntaxWarning: invalid escape sequence '\.'
2026-06-17T19:50:10.862420685+02:00 stderr F   """Configuration class for applying different quantization configs to modules or parameters based on their fully qualified names (FQNs).
2026-06-17T19:50:16.111736697+02:00 stderr F [2026-06-17 17:50:16] Tokenizer for /data/hub/models--lukealonso--GLM-5.2-NVFP4/snapshots/8a1f4a13204acf2b7ac840375efaed64c231c522 is still TokenizersBackend after retries with --trust-remote-code. Model-specific tokenizer attributes may be missing.
2026-06-17T19:50:17.70943276+02:00 stderr F [transformers] The following generation flags are not valid and may be ignored: ['top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
2026-06-17T19:50:18.755571126+02:00 stderr F [transformers] The following generation flags are not valid and may be ignored: ['top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
2026-06-17T19:50:23.350800698+02:00 stderr F [2026-06-17 17:50:23 PP0] Tokenizer for /data/hub/models--lukealonso--GLM-5.2-NVFP4/snapshots/8a1f4a13204acf2b7ac840375efaed64c231c522 is still TokenizersBackend after retries with --trust-remote-code. Model-specific tokenizer attributes may be missing.
2026-06-17T19:50:23.974409369+02:00 stderr F [2026-06-17 17:50:23 PP1] Tokenizer for /data/hub/models--lukealonso--GLM-5.2-NVFP4/snapshots/8a1f4a13204acf2b7ac840375efaed64c231c522 is still TokenizersBackend after retries with --trust-remote-code. Model-specific tokenizer attributes may be missing.
2026-06-17T19:50:24.295424018+02:00 stderr F [2026-06-17 17:50:24 PP0] Init torch distributed begin.
2026-06-17T19:50:24.752578767+02:00 stderr F [2026-06-17 17:50:24 PP1] Init torch distributed begin.
2026-06-17T19:50:25.44513692+02:00 stderr F [2026-06-17 17:50:25 PP0] sglang is using nccl==2.28.9
2026-06-17T19:50:26.948140626+02:00 stderr F [2026-06-17 17:50:26 PP0] Init torch distributed ends. elapsed=2.65 s, mem usage=1.13 GB
2026-06-17T19:50:26.948181402+02:00 stderr F [2026-06-17 17:50:26 PP1] Init torch distributed ends. elapsed=2.20 s, mem usage=1.13 GB
2026-06-17T19:50:30.873863844+02:00 stderr F [2026-06-17 17:50:30 PP0] Load weight begin. avail mem=265.90 GB
2026-06-17T19:50:30.874317913+02:00 stderr F [2026-06-17 17:50:30 PP0] Using ModelOptModelLoader due to ModelOpt quantization config.
2026-06-17T19:50:30.87440252+02:00 stderr F [2026-06-17 17:50:30 PP0] ModelOptModelLoader: Loading base model...
2026-06-17T19:50:30.87443714+02:00 stderr F [2026-06-17 17:50:30 PP0] Model is already quantized, loading directly...
2026-06-17T19:50:30.874583166+02:00 stderr F [2026-06-17 17:50:30 PP0] Detected nvfp4 checkpoint. Please note that the format is experimental and subject to change.
2026-06-17T19:50:31.188518826+02:00 stderr F [2026-06-17 17:50:31 PP0] FlashInfer TRTLLM MoE deferred finalize is disabled (moe_runner_backend=flashinfer_trtllm, quant_method=ModelOptNvFp4FusedMoEMethod).
2026-06-17T19:50:31.190006823+02:00 stderr F [2026-06-17 17:50:31 PP0] Setting inplace to False for FlashInfer TRTLLM MoE backend.
Multi-thread loading shards:   0% Completed | 0/87 [00:00<?, ?it/s][2026-06-17 17:50:31 PP0] lm_head.weight not found in params_dict.
2026-06-17T19:50:32.059286578+02:00 stderr F [2026-06-17 17:50:32 PP1] Load weight begin. avail mem=265.90 GB
2026-06-17T19:50:32.059650059+02:00 stderr F [2026-06-17 17:50:32 PP1] Using ModelOptModelLoader due to ModelOpt quantization config.
2026-06-17T19:50:32.05969317+02:00 stderr F [2026-06-17 17:50:32 PP1] ModelOptModelLoader: Loading base model...
2026-06-17T19:50:32.059722026+02:00 stderr F [2026-06-17 17:50:32 PP1] Model is already quantized, loading directly...
2026-06-17T19:50:32.059851312+02:00 stderr F [2026-06-17 17:50:32 PP1] Detected nvfp4 checkpoint. Please note that the format is experimental and subject to change.
Multi-thread loading shards:   1% Completed | 1/87 [00:00<00:19,  4.47it/s][2026-06-17 17:50:32 PP1] FlashInfer TRTLLM MoE deferred finalize is disabled (moe_runner_backend=flashinfer_trtllm, quant_method=ModelOptNvFp4FusedMoEMethod).
2026-06-17T19:50:32.463936189+02:00 stderr F [2026-06-17 17:50:32 PP1] Setting inplace to False for FlashInfer TRTLLM MoE backend.
Multi-thread loading shards: 100% Completed | 87/87 [00:31<00:00,  2.79it/s]
2026-06-17T19:51:40.807190453+02:00 stderr F [2026-06-17 17:51:40 PP0] Load weight end. elapsed=69.93 s, type=GlmMoeDsaForCausalLM, quant=modelopt_fp4, quant_algo=NVFP4, avail mem=63.89 GB, mem usage=202.01 GB.
2026-06-17T19:51:44.269900016+02:00 stderr F [2026-06-17 17:51:44 PP1] Load weight end. elapsed=72.21 s, type=GlmMoeDsaForCausalLM, quant=modelopt_fp4, quant_algo=NVFP4, avail mem=49.76 GB, mem usage=216.14 GB.
2026-06-17T19:51:48.218258042+02:00 stderr F [2026-06-17 17:51:48 PP0] Tokenizer for /data/hub/models--lukealonso--GLM-5.2-NVFP4/snapshots/8a1f4a13204acf2b7ac840375efaed64c231c522 is still TokenizersBackend after retries with --trust-remote-code. Model-specific tokenizer attributes may be missing.
2026-06-17T19:51:49.931918695+02:00 stderr F [2026-06-17 17:51:49 PP1] Tokenizer for /data/hub/models--lukealonso--GLM-5.2-NVFP4/snapshots/8a1f4a13204acf2b7ac840375efaed64c231c522 is still TokenizersBackend after retries with --trust-remote-code. Model-specific tokenizer attributes may be missing.
2026-06-17T19:51:50.651923729+02:00 stderr F [2026-06-17 17:51:50 PP0] Max concurrent requests (per dp worker) from the finalized token capacity: max_num_reqs=2048.
2026-06-17T19:51:50.651972386+02:00 stderr F [2026-06-17 17:51:50 PP1] Max concurrent requests (per dp worker) from the finalized token capacity: max_num_reqs=2048.
2026-06-17T19:51:50.694148737+02:00 stderr F [2026-06-17 17:51:50 PP1] KV Cache is allocated. dtype: torch.bfloat16, #tokens: 747520, KV size: 34.87 GB
2026-06-17T19:51:50.69454151+02:00 stderr F [2026-06-17 17:51:50 PP1] Memory pool end. avail mem=6.79 GB
2026-06-17T19:51:50.696040671+02:00 stderr F [2026-06-17 17:51:50 PP0] KV Cache is allocated. dtype: torch.bfloat16, #tokens: 747520, KV size: 34.87 GB
2026-06-17T19:51:50.696521383+02:00 stderr F [2026-06-17 17:51:50 PP0] Memory pool end. avail mem=20.92 GB
2026-06-17T19:51:51.01932207+02:00 stderr F [2026-06-17 17:51:51 PP1] Running FlashInfer autotune with cache: /data/glm-5.2/.cache/sglang/flashinfer/autotune/0.6.12/sm103/40c42f57c7f02ba3/rank_tp0_pp1_dp0.json
2026-06-17T19:51:51.160723894+02:00 stderr F [2026-06-17 17:51:51 PP0] Running FlashInfer autotune with cache: /data/glm-5.2/.cache/sglang/flashinfer/autotune/0.6.12/sm103/40c42f57c7f02ba3/rank_tp0_pp0_dp0.json
2026-06-17T19:51:53.290513721+02:00 stderr F [2026-06-17 17:51:53 PP1] Scheduler hit an exception: Traceback (most recent call last):
2026-06-17T19:51:53.29054134+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 4113, in run_scheduler_process
2026-06-17T19:51:53.290561479+02:00 stderr F     scheduler = Scheduler(
2026-06-17T19:51:53.290563049+02:00 stderr F                 ^^^^^^^^^^
2026-06-17T19:51:53.290563685+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 432, in __init__
2026-06-17T19:51:53.290564367+02:00 stderr F     self.init_model_worker()
2026-06-17T19:51:53.290564997+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 863, in init_model_worker
2026-06-17T19:51:53.290565637+02:00 stderr F     self.init_all_backends()
2026-06-17T19:51:53.290566287+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 845, in init_all_backends
2026-06-17T19:51:53.290567717+02:00 stderr F     self.tp_worker.init_backends()
2026-06-17T19:51:53.290568709+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/managers/tp_worker.py", line 338, in init_backends
2026-06-17T19:51:53.290569511+02:00 stderr F     self.model_runner.init_backends(disable_cuda_graph=disable_cuda_graph)
2026-06-17T19:51:53.290570686+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 898, in init_backends
2026-06-17T19:51:53.290572215+02:00 stderr F     self.kernel_warmup()
2026-06-17T19:51:53.29057317+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 2487, in kernel_warmup
2026-06-17T19:51:53.290574886+02:00 stderr F     self._flashinfer_autotune()
2026-06-17T19:51:53.290576273+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 2605, in _flashinfer_autotune
2026-06-17T19:51:53.290577375+02:00 stderr F     self._dummy_run(batch_size=self.req_to_token_pool.size)
2026-06-17T19:51:53.290578281+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 2886, in _dummy_run
2026-06-17T19:51:53.290578972+02:00 stderr F     run_once()
2026-06-17T19:51:53.290579908+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 2874, in run_once
2026-06-17T19:51:53.290580712+02:00 stderr F     logits_output_or_pp_proxy_tensors = self.model.forward(
2026-06-17T19:51:53.290581326+02:00 stderr F                                         ^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290582216+02:00 stderr F   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
2026-06-17T19:51:53.290624738+02:00 stderr F     return func(*args, **kwargs)
2026-06-17T19:51:53.290626074+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290627165+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2736, in forward
2026-06-17T19:51:53.29062815+02:00 stderr F     hidden_states = self.model(
2026-06-17T19:51:53.29062903+02:00 stderr F                     ^^^^^^^^^^^
2026-06-17T19:51:53.290629896+02:00 stderr F   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1779, in _wrapped_call_impl
2026-06-17T19:51:53.290630735+02:00 stderr F     return self._call_impl(*args, **kwargs)
2026-06-17T19:51:53.290631608+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290632449+02:00 stderr F   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1790, in _call_impl
2026-06-17T19:51:53.290633304+02:00 stderr F     return forward_call(*args, **kwargs)
2026-06-17T19:51:53.290634541+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290635603+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2495, in forward
2026-06-17T19:51:53.290636852+02:00 stderr F     hidden_states, residual, topk_indices = layer(
2026-06-17T19:51:53.290637826+02:00 stderr F                                             ^^^^^^
2026-06-17T19:51:53.290638682+02:00 stderr F   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1779, in _wrapped_call_impl
2026-06-17T19:51:53.290643766+02:00 stderr F     return self._call_impl(*args, **kwargs)
2026-06-17T19:51:53.290647558+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290648263+02:00 stderr F   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1790, in _call_impl
2026-06-17T19:51:53.290648945+02:00 stderr F     return forward_call(*args, **kwargs)
2026-06-17T19:51:53.290649603+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290650234+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2142, in forward
2026-06-17T19:51:53.290650913+02:00 stderr F     hidden_states = self.self_attn(
2026-06-17T19:51:53.290651518+02:00 stderr F                     ^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290652135+02:00 stderr F   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1779, in _wrapped_call_impl
2026-06-17T19:51:53.290652756+02:00 stderr F     return self._call_impl(*args, **kwargs)
2026-06-17T19:51:53.290653358+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290654026+02:00 stderr F   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1790, in _call_impl
2026-06-17T19:51:53.290654637+02:00 stderr F     return forward_call(*args, **kwargs)
2026-06-17T19:51:53.290657017+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.29065792+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 1797, in forward
2026-06-17T19:51:53.290658569+02:00 stderr F     return self.forward_core(s)
2026-06-17T19:51:53.290659186+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290659808+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 1908, in forward_core
2026-06-17T19:51:53.290660424+02:00 stderr F     return self.forward_absorb_core(*inner_state)
2026-06-17T19:51:53.2906611+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290662325+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py", line 530, in forward_absorb_core
2026-06-17T19:51:53.290663073+02:00 stderr F     attn_output = self.attn_mqa(
2026-06-17T19:51:53.290663707+02:00 stderr F                   ^^^^^^^^^^^^^^
2026-06-17T19:51:53.290675912+02:00 stderr F   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1779, in _wrapped_call_impl
2026-06-17T19:51:53.290676568+02:00 stderr F     return self._call_impl(*args, **kwargs)
2026-06-17T19:51:53.290677192+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290678979+02:00 stderr F   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1790, in _call_impl
2026-06-17T19:51:53.290679904+02:00 stderr F     return forward_call(*args, **kwargs)
2026-06-17T19:51:53.290680504+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290681134+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/layers/radix_attention.py", line 145, in forward
2026-06-17T19:51:53.29068174+02:00 stderr F     return get_attn_backend().forward(
2026-06-17T19:51:53.290682591+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290683293+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/layers/attention/base_attn_backend.py", line 169, in forward
2026-06-17T19:51:53.290684006+02:00 stderr F     return self.forward_decode(
2026-06-17T19:51:53.290684946+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290685748+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/layers/attention/dsa_backend.py", line 1587, in forward_decode
2026-06-17T19:51:53.29068638+02:00 stderr F     return self._forward_trtllm(
2026-06-17T19:51:53.290686998+02:00 stderr F            ^^^^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290690463+02:00 stderr F   File "/sgl-workspace/sglang/python/sglang/srt/layers/attention/dsa_backend.py", line 2178, in _forward_trtllm
2026-06-17T19:51:53.290691092+02:00 stderr F     batch_size = page_table_1.shape[0]
2026-06-17T19:51:53.290691833+02:00 stderr F                  ^^^^^^^^^^^^^^^^^^
2026-06-17T19:51:53.290692479+02:00 stderr F AttributeError: 'NoneType' object has no attribute 'shape'
2026-06-17T19:51:53.290693203+02:00 stderr F
2026-06-17T19:51:53.291020372+02:00 stderr F [2026-06-17 17:51:53] Received sigquit from a child process. It usually means the child failed.
2026-06-17T19:51:53.291100201+02:00 stderr F [2026-06-17 17:51:53] kill_process_tree called: parent_pid=1, include_parent=True, pid=1
2026-06-17T19:51:53.30397646+02:00 stderr F [2026-06-17 17:51:53] kill_process_tree called: parent_pid=1, include_parent=False, pid=1
```

### Reproduction

Run the command above.

### Environment

```
Python: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0]
CUDA available: True
GPU 0,1: NVIDIA B300 SXM6 AC
GPU 0,1 Compute Capability: 10.3
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 595.71.05
PyTorch: 2.11.0+cu130
sglang: 0.0.0.dev1+g0d651e653
sglang-kernel: 0.4.3
flashinfer_python: 0.6.12
flashinfer_cubin: 0.6.12
flashinfer_jit_cache: 0.6.12+cu130
triton: 3.6.0
transformers: 5.8.1
torchao: 0.17.0+cu130
numpy: 2.3.5
aiohttp: 3.14.1
fastapi: 0.137.1
huggingface_hub: 1.19.0
interegular: 0.3.3
modelscope: 1.37.1
orjson: 3.11.9
outlines: 0.1.11
packaging: 26.2
psutil: 7.2.2
pydantic: 2.13.4
python-multipart: 0.0.32
pyzmq: 27.1.0
uvicorn: 0.49.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.13.0
anthropic: 0.109.2
litellm: Module Not Found
torchcodec: 0.11.1+cu130
NVIDIA Topology: 
        GPU0    GPU1    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    NIC6    NIC7    NIC8    NIC9    NIC10   NIC11   NIC12   NIC13   NIC14   NIC15   NIC16   NIC17   NIC18   NIC19   NIC20   NIC21   NIC22   NIC23   CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV18    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    PXB     PXB     NODE    NODE    64-127,192-255  1               N/A
GPU1    NV18     X      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    PXB     PXB     64-127,192-255  1               N/A
NIC0    SYS     SYS      X      PIX     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC1    SYS     SYS     PIX      X      NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC2    SYS     SYS     NODE    NODE     X      PIX     PIX     PIX     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC3    SYS     SYS     NODE    NODE    PIX      X      PIX     PIX     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC4    SYS     SYS     NODE    NODE    PIX     PIX      X      PIX     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC5    SYS     SYS     NODE    NODE    PIX     PIX     PIX      X      NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC6    SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE     X      PIX     NODE    NODE    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC7    SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    PIX      X      NODE    NODE    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC8    SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE     X      PIX     NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC9    SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    PIX      X      NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC10   SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE     X      PIX     NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC11   SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    PIX      X      NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC12   SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE     X      PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC13   SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    PIX      X      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC14   NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      PIX     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE
NIC15   NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     PIX      X      NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE
NIC16   NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE     X      PIX     NODE    NODE    NODE    NODE    NODE    NODE
NIC17   NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    PIX      X      NODE    NODE    NODE    NODE    NODE    NODE
NIC18   NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE     X      PIX     NODE    NODE    NODE    NODE
NIC19   NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE    PIX      X      NODE    NODE    NODE    NODE
NIC20   PXB     NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE     X      PIX     NODE    NODE
NIC21   PXB     NODE    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    PIX      X      NODE    NODE
NIC22   NODE    PXB     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE     X      PIX
NIC23   NODE    PXB     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    NODE    NODE    PIX      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1
  NIC2: mlx5_2
  NIC3: mlx5_3
  NIC4: mlx5_4
  NIC5: mlx5_5
  NIC6: mlx5_6
  NIC7: mlx5_7
  NIC8: mlx5_8
  NIC9: mlx5_9
  NIC10: mlx5_10
  NIC11: mlx5_11
  NIC12: mlx5_12
  NIC13: mlx5_13
  NIC14: mlx5_14
  NIC15: mlx5_15
  NIC16: mlx5_16
  NIC17: mlx5_17
  NIC18: mlx5_18
  NIC19: mlx5_19
  NIC20: mlx5_20
  NIC21: mlx5_21
  NIC22: mlx5_22
  NIC23: mlx5_23


ulimit soft: 1048576
```

@mmangkad still no:
```
2026-06-20 11:33:56 PP1] Running FlashInfer autotune with cache: /data/glm-5.2/.cache/sglang/flashinfer/autotune/0.6.12/sm103/a8f31462fb6419a4/rank_tp0_pp1_dp0.json
[2026-06-20 11:33:56 PP1] Scheduler hit an exception: Traceback (most recent call last):
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 4170, in run_scheduler_process
    scheduler = Scheduler(
                ^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 432, in __init__
    self.init_model_worker()
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 863, in init_model_worker
    self.init_all_backends()
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 845, in init_all_backends
    self.tp_worker.init_backends()
  File "/sgl-workspace/sglang/python/sglang/srt/managers/tp_worker.py", line 338, in init_backends
    self.model_runner.init_backends(disable_cuda_graph=disable_cuda_graph)
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 909, in init_backends
    self.eager_runner = EagerRunner(self)
                        ^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/runner/eager_runner.py", line 133, in __init__
    self.warmup()
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/runner/base_runner.py", line 216, in warmup
    self._flashinfer_autotune(buffers=buffers, batch_size=batch_size)
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/runner/base_runner.py", line 358, in _flashinfer_autotune
    self._dummy_run(batch_size=batch_size, buffers=buffers)
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/runner/base_runner.py", line 696, in _dummy_run
    run_once()
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/runner/base_runner.py", line 684, in run_once
    logits_output_or_pp_proxy_tensors = mr.model.forward(
                                        ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2745, in forward
    hidden_states = self.model(
                    ^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1779, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1790, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2410, in forward
    assert not (
           ^^^^^
AssertionError: PP stage starting at layer 19 requires DSA topk_indices from the previous stage.
[2026-06-20 11:33:56] Received sigquit from a child process. It usually means the child failed.
[2026-06-20 11:33:56] kill_process_tree called: parent_pid=1, include_parent=True, pid=1
[2026-06-20 11:33:56] kill_process_tree called: parent_pid=1, include_parent=False, pid=1
```

One step closer. This time it at least starts, but then:
```
[2026-06-20 18:38:13 PP0] Scheduler hit an exception: Traceback (most recent call last):
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 4186, in run_scheduler_process
    scheduler.run_event_loop()
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 1470, in run_event_loop
    dispatch_event_loop(self)
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 4047, in dispatch_event_loop
    scheduler.event_loop_pp()
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler_pp_mixin.py", line 128, in event_loop_pp
    result, self.launch_event = self._pp_launch_batch(
                                ^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler_pp_mixin.py", line 1247, in _pp_launch_batch
    result = self.run_batch(self.cur_batch, pp_proxy_tensors)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/utils/nvtx_utils.py", line 109, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 3234, in run_batch
    batch_result = self.model_worker.forward_batch_generation(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/managers/tp_worker.py", line 548, in forward_batch_generation
    out = self.model_runner.forward(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 2922, in forward
    output = self._forward_raw(
             ^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 3048, in _forward_raw
    ret = self.eager_runner.execute(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/runner/eager_runner.py", line 256, in execute
    return self._execute_extend(forward_batch, pp_proxy_tensors)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/runner/eager_runner.py", line 404, in _execute_extend
    ret = model_runner.model.forward(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2745, in forward
    hidden_states = self.model(
                    ^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1779, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1790, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2525, in forward
    assert topk_indices is not None, (
           ^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: PP stage ending at layer 19 must forward DSA topk_indices because the next stage starts on a skip-topk layer.
[2026-06-20 18:38:13] SIGQUIT received. signum=None, frame=None. It usually means one child failed.
```

> One step closer. This time it at least starts, but then:
> 
> ```
> [2026-06-20 18:38:13 PP0] Scheduler hit an exception: Traceback (most recent call last):
>   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 4186, in run_scheduler_process
>     scheduler.run_event_loop()
>   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 1470, in run_event_loop
>     dispatch_event_loop(self)
>   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 4047, in dispatch_event_loop
>     scheduler.event_loop_pp()
>   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
>     return func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler_pp_mixin.py", line 128, in event_loop_pp
>     result, self.launch_event = self._pp_launch_batch(
>                                 ^^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler_pp_mixin.py", line 1247, in _pp_launch_batch
>     result = self.run_batch(self.cur_batch, pp_proxy_tensors)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/utils/nvtx_utils.py", line 109, in wrapper
>     return func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/managers/scheduler.py", line 3234, in run_batch
>     batch_result = self.model_worker.forward_batch_generation(
>                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/managers/tp_worker.py", line 548, in forward_batch_generation
>     out = self.model_runner.forward(
>           ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 2922, in forward
>     output = self._forward_raw(
>              ^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 3048, in _forward_raw
>     ret = self.eager_runner.execute(
>           ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/model_executor/runner/eager_runner.py", line 256, in execute
>     return self._execute_extend(forward_batch, pp_proxy_tensors)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/model_executor/runner/eager_runner.py", line 404, in _execute_extend
>     ret = model_runner.model.forward(
>           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 124, in decorate_context
>     return func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2745, in forward
>     hidden_states = self.model(
>                     ^^^^^^^^^^^
>   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1779, in _wrapped_call_impl
>     return self._call_impl(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py", line 1790, in _call_impl
>     return forward_call(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/sgl-workspace/sglang/python/sglang/srt/models/deepseek_v2.py", line 2525, in forward
>     assert topk_indices is not None, (
>            ^^^^^^^^^^^^^^^^^^^^^^^^
> AssertionError: PP stage ending at layer 19 must forward DSA topk_indices because the next stage starts on a skip-topk layer.
> [2026-06-20 18:38:13] SIGQUIT received. signum=None, frame=None. It usually means one child failed.
> ```

@xhejtman Could you try `export SGLANG_PP_LAYER_PARTITION=18,20,20,20`?
I am feeling like this kind of topk layer skipping issues both could be simply addressed by a smarter layer partition strategy.

--- a/python/sglang/srt/model_executor/runner/base_runner.py
+++ b/python/sglang/srt/model_executor/runner/base_runner.py
@@ -78,6 +78,7 @@ def _allocate_decode_buffers(
     enable_mamba_track: bool,
     ne_token_table: Optional[torch.Tensor] = None,
     hc_hidden_size: Optional[int] = None,
+    pp_proxy_topk_size: Optional[int] = None,
 ) -> SimpleNamespace:
     """Allocate the FB-shared decode buffers."""
     with torch.device(device):
@@ -114,6 +115,10 @@ def _allocate_decode_buffers(
             if not is_mhc:
                 pp_proxy_tensors["residual"] = torch.zeros(
                     (max_bs, hidden_size), dtype=dtype
+                )
+            if pp_proxy_topk_size is not None:
+                pp_proxy_tensors["topk_indices"] = torch.zeros(
+                    (max_num_token, pp_proxy_topk_size), dtype=torch.int32
                 )
         else:
             pp_proxy_tensors = None
@@ -429,6 +434,7 @@ class BaseRunner(ABC):
             cache_loc_dtype=torch.int64,
             enable_mamba_track=False,
             hc_hidden_size=getattr(mr.model_config, "hc_hidden_size", None),
+            pp_proxy_topk_size=mr.get_pp_proxy_topk_size(),
         )
 
     def _dummy_run(

--- a/python/sglang/srt/model_executor/runner/eager_runner.py
+++ b/python/sglang/srt/model_executor/runner/eager_runner.py
@@ -190,6 +190,11 @@ class EagerRunner(BaseRunner):
                 pp_proxy_tensors["residual"] = torch.zeros(
                     (rows, hidden_size), dtype=mr.dtype, device=mr.device
                 )
+            pp_proxy_topk_size = mr.get_pp_proxy_topk_size()
+            if pp_proxy_topk_size is not None:
+                pp_proxy_tensors["topk_indices"] = torch.zeros(
+                    (rows, pp_proxy_topk_size), dtype=torch.int32, device=mr.device
+                )
 
         adapter = SimpleNamespace(
             input_ids=_slot("input_ids"),

--- a/python/sglang/srt/models/deepseek_v2.py
+++ b/python/sglang/srt/models/deepseek_v2.py
@@ -2387,6 +2387,13 @@ class DeepseekV2Model(nn.Module):
     def get_input_embeddings(self) -> torch.Tensor:
         return self.embed_tokens
 
+    def _dsa_forward_uses_topk(self) -> bool:
+        if not self.use_dsa:
+            return False
+        backend = get_attn_backend()
+        backend = getattr(backend, "primary", backend)
+        return not getattr(backend, "use_mha", False)
+
     def forward(
         self,
         input_ids: torch.Tensor,
@@ -2396,6 +2403,7 @@ class DeepseekV2Model(nn.Module):
         pp_proxy_tensors: Optional[PPProxyTensors] = None,
     ) -> Union[torch.Tensor, PPProxyTensors]:
         total_num_layers = self.end_layer - self.start_layer
+        dsa_forward_uses_topk = self._dsa_forward_uses_topk()
         if self.pp_group.is_first_rank:
             if input_embeds is None:
                 hidden_states = self.embed_tokens(input_ids)
@@ -2411,6 +2419,7 @@ class DeepseekV2Model(nn.Module):
                 not forward_batch.forward_mode.is_idle()
                 and hidden_states.shape[0] != 0
                 and self.use_dsa
+                and dsa_forward_uses_topk
                 and dsa_layer_skips_topk(self.config, self.start_layer)
                 and topk_indices is None
             ), (
@@ -2515,6 +2524,7 @@ class DeepseekV2Model(nn.Module):
             }
             if (
                 self.use_dsa
+                and dsa_forward_uses_topk
                 and self.end_layer < self.config.num_hidden_layers
                 and dsa_layer_skips_topk(self.config, self.end_layer)
             ):

